### DISCUSSION TITLE
YOLOv5n Nano Study

### DISCUSSION THREAD

**Discussion by glenn-jocher** [ultralytics/yolov5#5027] [Ideas]
Starting this discussion to track smaller model experiments as several users have expressed interest. Experiments in https://github.com/ultralytics/yolov5/tree/tests/v6.0. Note: only inference speeds shown, NMS not included in charts below.

### Train

```
python train.py --cfg yolov5n1.yaml
python train.py --cfg yolov5n2.yaml
python train.py --cfg yolov5n3.yaml
python train.py --cfg yolov5n4.yaml
python train.py --cfg yolov5s_v6.yaml
```

### Val: V100 `--batch 32`

```bash
python val.py --batch 32 --data coco.yaml --task speed --weights runs/train/exp/weights/best.pt runs/train/exp2/weights/best.pt runs/train/exp3/weights/best.pt runs/train/exp4/weights/best.pt runs/train/exp5/weights/best.pt yolov5s.pt
```

|Model |size<br><sup>(pixels) |mAP<sup>val<br>0.5:0.95 |mAP<sup>test<br>0.5:0.95 |mAP<sup>val<br>0.5 |Speed<br><sup>V100 (ms)| |params<br><sup>(M) |FLOPs<br><sup>640 (B)
|---                    |---  |---      |---      |---      |---     |---|---   |---
|YOLOv5n1      |640  |26.1     |-     |41.9     |**0.9** |   |1.9   |4.5
|YOLOv5n2      |640  |24.1     |-     |39.8     |**0.7** |   |1.3   |3.3
|YOLOv5n3      |640  |22.9     |-     |38.4     |**0.7** |   |1.4   |3.2
|YOLOv5n4      |640  |27.6     |-     |50.1     |**0.9** |   |2.9   |6.7
|                       |     |         |         |         |        |   |      |
|YOLOv5s release 5.0 (current)     |640  |36.7     |-     |55.4     |**1.0** |   |7.3   |17.0
|YOLOv5s release 6.0 (pending)      |640  |37.2     |-     |56.0     |**1.0** |   |7.2   |16.5

### Val: V100 `--batch 1`

```bash
python val.py --batch 1 --data coco.yaml --task speed --weights runs/train/exp/weights/best.pt runs/train/exp2/weights/best.pt runs/train/exp3/weights/best.pt runs/train/exp4/weights/best.pt runs/train/exp5/weights/best.pt yolov5s.pt
```

|Model |size<br><sup>(pixels) |mAP<sup>val<br>0.5:0.95 |mAP<sup>test<br>0.5:0.95 |mAP<sup>val<br>0.5 |Speed<br><sup>V100 (ms)| |params<br><sup>(M) |FLOPs<br><sup>640 (B)
|---                    |---  |---      |---      |---      |---     |---|---   |---
|YOLOv5n1      |640  |-     |-     |-     |**7.8** |   |1.9   |4.5
|YOLOv5n2      |640  |-     |-     |-     |**7.4** |   |1.3   |3.3
|YOLOv5n3      |640  |-     |-     |-     |**6.9** |   |1.4   |3.2
|YOLOv5n4      |640  |-     |-     |-     |**6.8** |   |2.9   |6.7
|                       |     |         |         |         |        |   |      |
|YOLOv5s release 5.0 (current)     |640  |36.7     |-     |55.4     |**8.2** |   |7.3   |17.0
|YOLOv5s release 6.0 (pending)      |640  |37.2     |-     |56.0     |**8.2** |   |7.2   |16.5

### Val: V100 `--batch 1 --device cpu`

```bash
python val.py --batch 1 --device cpu --data coco128.yaml --task speed --weights runs/train/exp/weights/best.pt runs/train/exp2/weights/best.pt runs/train/exp3/weights/best.pt runs/train/exp4/weights/best.pt runs/train/exp5/weights/best.pt yolov5s.pt
```

|Model |size<br><sup>(pixels) |mAP<sup>val<br>0.5:0.95 |mAP<sup>test<br>0.5:0.95 |mAP<sup>val<br>0.5 |Speed<br><sup>CPU (ms)| |params<br><sup>(M) |FLOPs<br><sup>640 (B)
|---                    |---  |---      |---      |---      |---     |---|---   |---
|YOLOv5n1      |640  |-     |-     |-     |**35.2** |   |1.9   |4.5
|YOLOv5n2      |640  |-     |-     |-     |**34.0** |   |1.3   |3.3
|YOLOv5n3      |640  |-     |-     |-     |**32.3** |   |1.4   |3.2
|YOLOv5n4      |640  |-     |-     |-     |**54.8** |   |2.9   |6.7
|                       |     |         |         |         |        |   |      |
|YOLOv5s release 5.0 (current)     |640  |36.7     |-     |55.4     |**83.5** |   |7.3   |17.0
|YOLOv5s release 6.0 (pending)      |640  |37.2     |-     |56.0     |**76.0** |   |7.2   |16.5

**Comment by ppogg** [comment]
Sir, if you have tested the speed of yolov5n on mobile devices? It seems that the Apple m1 chip will speed up the model inference, but how about the performance of yolov5n for ordinary mobile devices, like Android?

By the way, good job！Sir！I like it！

**Reply by glenn-jocher** [reply]
@ppogg we have not deployed v6.0 to iOS yet, we will do this in the next few days. The nano models should see the most speedup on CPU and low power mobile devices compared to the small models.

**Reply by ppogg** [reply]
Great, sir. I look forward to your work progress！

**Comment by glenn-jocher** [comment]
UDPATE: The result of the above study is that the **YOLOv5n2** model was selected as the best speed-mAP compromise candidate of the four experimental nano models. So **the v6.0 release YOLOv5n model is the YOLOv5n2 model from this study**. 

This model was also the simplest to understand as we only apply a reduction in width scaling from YOLOv5s (0.5) to YOLOv5n (0.25), which is in line with the scaling of the entire family (0.25 changes between each size).

**Comment by henbucuoshanghai** [comment]
the nano model of yolov5-6.0 can be 30fps in 4G Nano? img is 640*640*3

**Reply by glenn-jocher** [reply]
@henbucuoshanghai I don't understand your question. The above profiling results are at `--img 640`

**Reply by njustczr** [reply]
是的 yolov5 n 在nano 4GB上 640x640x3的输入能达到30fps+

**Comment by henbucuoshanghai** [comment]
tks for reply.&nbsp;&nbsp; I mean the fps of the yolov5-6.0 on NVIDIA JETSON NANO?

**Reply by njustczr** [reply]
yolov5n-5.0 on NVIDIA JETSON NANO, the input size is 640x640x3,  i can get 30 fps+.

**Comment by danurrahmajati** [comment]
halo,  since the focus  on mobile use, is there any example of your implementation to the android device?

**Reply by kurtgenc** [reply]
Take a look at pytorch web there are many examples:
https://pytorch.org/mobile/android/

**Reply by danurrahmajati** [reply]
great thanks

**Comment by kurtgenc** [comment]
I have a custom model trained with yolov5l that I'd like to test inference on Android, is there a way to convert the trained model from yolov5l.pt to yolov5s.pt or other nano models? Or do I need to train for the smaller weight again?

**Reply by senstar-hsoleimani** [reply]
Train it again

**Comment by grobomacOakD** [comment]
Is it possible to convert a trained yolov5s to yolov5n?

**Reply by glenn-jocher** [reply]
@grobomacOakD haha, no. This would be like converting a dog brain into a cat brain.

What you can do is start training a YOLOv5n model from a pretrained YOLOv5s model, though very few layers will transfer:
```bash
python train.py --weights yolov5s.pt --cfg yolov5n.yaml
...
Transferred 60/349 items from yolov5s.pt
```

You're probably just better of training a new YOLOv5n model starting from our pretrained model:
```bash
python train.py --weights yolov5n.pt
```

**Comment by henningscale** [comment]
Why are these models not run as .engine or .onnx files?
Will the .pt files get published for us to try? I cannot find them manually or with model = torch.hub.load('ultralytics/yolov5', 'yolov5n1') or is it the case that Yolov5n is the same as Yolov5n2?

**Reply by glenn-jocher** [reply]
@henningscale yes, result of this study was selection of n2 as the YOLOv5n model for the v6.0 release.

**Reply by henningscale** [reply]
@glenn-jocher How do you get inference speeds that quick? With my 4gb nano and using the v5n6 I get 61ms inference speeds average from 210 images with: python3 detect.py --weights yolov5n6 --nosave --img640 640 --source .../data.

**Reply by glenn-jocher** [reply]
@henningscale commands and hardware to reproduce are clearly indicated in my original post https://github.com/ultralytics/yolov5/discussions/5027#discussion-3603985, which have zero in common with your hardware and commands. If you compare apples to oranges you shouldn't be surprised to find differences.

<img width="895" alt="Screen Shot 2022-03-08 at 12 06 55 PM" src="https://user-images.githubusercontent.com/26833433/157226666-6377640c-0f12-41de-89c2-98b133ab0ad8.png">

**Comment by Sohna-Ctrl** [comment]
can you share the config file of the yolov5n1、yolov5n2、yolov5n3、yolov5n4，the website(https://github.com/ultralytics/yolov5/tree/tests/v6.0.) is disabled

**Reply by glenn-jocher** [reply]
@Sohna-Ctrl oh sorry, it seems the branch was auto-deleted upon PR merge. I think you can rest assured that the current [yolov5n.yaml](https://github.com/ultralytics/yolov5/blob/master/models/yolov5n.yaml) is the best balance of speed and accuracy though among the ones tested.

**Reply by Sohna-Ctrl** [reply]
thanks！but now i have some project needs，maybe your config setting can inspire me to change the model  to better fit my datasets.

**Reply by Sohna-Ctrl** [reply]
Can you re-upload the files for me to study，if it is convenient.

**Comment by R97416032** [comment]
Hello,sir.Can you tell me how to compute the size of model? The params number of v5n is 1.9M but why the size is about 4MB？

**Reply by glenn-jocher** [reply]
Size is a function of quantization. FP16 models are 4MB.

**Comment by R97416032** [comment]
hi，sir！I get into trouble again. After training，it generates a results.txt. There are lr0、lr1 and lr2 in results.txt. I guess lr is learning rate, but I don't  know why it has three lr. Thank you sincerely!

**Reply by glenn-jocher** [reply]
@R97416032 YOLOv5 has 3 parameter groups defined here:
https://github.com/ultralytics/yolov5/blob/b24a576823a337357dfe3a1b705ab5459ebe8079/train.py#L153-L175

**Comment by R97416032** [comment]
genius!

**Comment by jjjonathan14** [comment]
Is it possible to get the architecture of yolov5n and pytorch implementation architecture from scratch

**Reply by glenn-jocher** [reply]
@jjjonathan14 👋 Hello! Thanks for asking about YOLOv5 🚀 **architecture visualization**. We've made visualizing YOLOv5 🚀 architectures super easy. There are 3 main ways:

## `model.yaml`

Each model has a corresponding yaml file that displays the model architecture. Here is YOLOv5s, defined by `yolov5s.yaml`:
https://github.com/ultralytics/yolov5/blob/1a3ecb8b386115fd22129eaf0760157b161efac7/models/yolov5s.yaml#L12-L48

## TensorBoard Graph

Simply start training a model, and then view the TensorBoard Graph for an interactive view of the model architecture. This example shows YOLOv5s viewed in our [Notebook](https://github.com/ultralytics/yolov5/blob/master/tutorial.ipynb) –  <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>

```python
# Tensorboard
%load_ext tensorboard
%tensorboard --logdir runs/train

# Train YOLOv5s on COCO128 for 3 epochs
python train.py --weights yolov5s.pt --epochs 3
```

<img width="893" alt="Screenshot 2021-04-11 at 01 10 09" src="https://user-images.githubusercontent.com/26833433/114286928-349bd600-9a63-11eb-941f-7139ee6cd602.png">

## Netron viewer

Use [https://netron.app](https://netron.app/) to view exported ONNX models:

```
python export.py --weights yolov5s.pt --include onnx --simplify
```

<img width="890" alt="Screen Shot 2022-04-29 at 11 09 23 AM" src="https://user-images.githubusercontent.com/26833433/165999628-0d6095ce-58dd-49b7-a04e-25d0995e5493.png">


Good luck 🍀 and let us know if you have any other questions!

**Comment by jjjonathan14** [comment]
Thank you so much for the quick reply Glenn Jocher and the support. Thanks
again
